File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.py
from collections import deque

def solution(queue1, queue2):
    q1 = deque(queue1)
    q2 = deque(queue2)
    
    # 두 큐의 합을 미리 계산
    sum1 = sum(q1)
    sum2 = sum(q2)
    total = sum1 + sum2
    
    # 총합이 홀수면 두 큐를 같게 만들 수 없음
    if total % 2 != 0:
        return -1
    
    # 목표로 하는 각 큐의 합
    target = total // 2
    
    cnt = 0
    limit = len(q1) + len(q2) * 2
    
    # 투포인터처럼 두 큐를 비교하면서 진행
    while sum1 != target and cnt <= limit:
        if sum1 > target:
            # q1의 합이 크면 q1에서 하나를 빼서 q2에 넣음
            value = q1.popleft()
            q2.append(value)
            sum1 -= value
        else:
            # q2의 합이 크면 q2에서 하나를 빼서 q1에 넣음
            value = q2.popleft()
            q1.append(value)
            sum2 -= value
            sum1 += value
        cnt += 1
    
    # 결과 확인
    if sum1 == target:
        return cnt
    else:
        return -1
    
    # 큐의 합을 미리 계산 안했다고 시간 초과가 날 줄은 몰랐다
    # 결론적으로 두 큐 중에 총 합이 더 큰 쪽에서 이동시키는 방법으로 해결 가능
    # 어떤 두 무언가의 합이 같아져야하는 문제에서 잘 생각하기

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        Queue<Integer> q1 = new ArrayDeque<>();
        Queue<Integer> q2 = new ArrayDeque<>();

        int len = queue1.length;
        long sum1 = 0, sum2 = 0;

        for (int i = 0; i < len; i++) {
            q1.offer(queue1[i]);
            sum1 += queue1[i];
            q2.offer(queue2[i]);
            sum2 += queue2[i];
        }

        if ((sum1 + sum2) % 2 == 1) return -1;

        long target = (sum1 + sum2) / 2;
        int answer = 0, cnt = 0;
        
        while (sum1 != target && cnt < len * 4) {
            if (sum1 > target) {
                int a = q1.poll();
                q2.offer(a);
                sum1 -= a;
            } else {
                int a = q2.poll();
                q1.offer(a);
                sum1 += a;
            }

            answer++;
            cnt++;
        }
    
        if (sum1 != target) return -1;
        return answer;
    }
}